### PR TITLE
Adds ledger client helper

### PR DESCRIPTION
**Pull request by NejcZdovc**
Resolves https://github.com/brave/brave-browser/issues/10981

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [ ] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:
N/A

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: vendor/bat-native-ledger/include/bat/ledger/ledger.h [hunk 1]
```diff
@@ -125,8 +125,6 @@ class LEDGER_EXPORT Ledger {
 
   virtual void OnTimer(uint32_t timer_id) = 0;
 
-  virtual std::string URIEncode(const std::string& value) = 0;
```
Comment (NejcZdovc) [inline vendor/bat-native-ledger/include/bat/ledger/ledger.h:128]: not used

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database_pending_contribution.cc [hunk 2]
```diff
@@ -67,7 +67,9 @@ void DatabasePendingContribution::InsertOrUpdateList(
       _1,
       callback);
 
-  ledger_->RunDBTransaction(std::move(transaction), transaction_callback);
+  ledger_->GetLedgerClient()->RunDBTransaction(
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/database/database_pending_contribution.cc:70]: Does it make more sense to pass a `LedgerClient` to some of these classes now instead of a `LedgerImpl`?

Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/database/database_pending_contribution.cc:70]: I would still go with ledger_impl as we have more moves like this so ledger_impl will hold all objects that you can use

Comment (emerick) [reply vendor/bat-native-ledger/src/bat/ledger/internal/database/database_pending_contribution.cc:70]: OK, just wanted to check, sounds good.

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc [hunk 3]
```diff
@@ -170,7 +171,23 @@ void DatabaseActivityInfo::NormalizeList(
       _1,
       callback);
 
-  ledger_->RunDBTransaction(std::move(transaction), transaction_callback);
+  auto shared_list = std::make_shared<ledger::PublisherInfoList>(
+      std::move(list));
+
+  ledger_->GetLedgerClient()->RunDBTransaction(
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc:177]: Is `transaction_callback` no longer used?

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc [hunk 4]
```diff
@@ -709,12 +715,6 @@ void LedgerImpl::ContributionCompleted(
   bat_contribution_->ContributionCompleted(
       result,
       contribution->Clone());
-
-  // TODO(https://github.com/brave/brave-browser/issues/7717)
-  // rename to ContributionCompleted
-  ledger_client_->OnReconcileComplete(
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc:715]: Is this removal intentional?

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc (hunk 3 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc:177]: removed

File: vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc (hunk 4 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/ledger_impl.cc:715]: yes, it was moved into `contribution.cc` - `Contribution::ContributionCompleted`

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.h [hunk 5]
```diff
@@ -23,7 +25,108 @@ class State {
...
+
+  std::vector<uint8_t> GetRecoverySeed();
+
+  void SetRecoverySeed(const std::vector<uint8_t>& seed);
+
+  std::string GetPaymentId();
+
+  void SetPaymentId(const std::string& id);
+
+  std::string ConvertInlineTipPlatformToString(
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/state/state.h:80]: Is this unused?

File: vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold.cc [hunk 6]
```diff
@@ -210,12 +211,15 @@ void Uphold::DisconnectWallet() {
...
-    [](ledger::Result _){});
+  ledger_->ledger_client()->ShowNotification(
+      "wallet_disconnected",
+      {},
+      [](ledger::Result _){});
 
-  ledger_->SaveExternalWallet(ledger::kWalletUphold, std::move(wallet));
-  ledger_->GetLedgerClient()->WalletDisconnected(ledger::kWalletUphold);
+  ledger_->ledger_client()->SaveExternalWallet(
+    ledger::kWalletUphold,
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold.cc:220]: nit: spacing

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc [hunk 7]
```diff
@@ -19,4 +87,319 @@ void State::Initialize(ledger::ResultCallback callback) {
...
+    ledger_->ResetReconcileStamp();
+    stamp = ledger_->ledger_client()->GetUint64State(
+        ledger::kStateNextReconcileStamp);
+  }
+
+  return stamp;
+}
+
+void State::SetReconcileStamp(
+    const int reconcile_interval) {
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc:204]: nit: could be on previous line

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc [hunk 8]
```diff
@@ -19,4 +87,319 @@ void State::Initialize(ledger::ResultCallback callback) {
...
+      enabled);
+}
+
+bool State::GetAutoContributeEnabled() {
+  return ledger_->ledger_client()->GetBooleanState(
+      ledger::kStateAutoContributeEnabled);
+}
+
+void State::SetAutoContributionAmount(
+    const double amount) {
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc:174]: nit: could be on previous line?

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc [hunk 9]
```diff
@@ -166,11 +167,23 @@ void DatabaseActivityInfo::NormalizeList(
...
+  ledger_->ledger_client()->RunDBTransaction(
+      std::move(transaction),
+      [this, shared_list, callback](ledger::DBCommandResponsePtr response) {
+        if (!response || response->status !=
+              ledger::DBCommandResponse::Status::RESPONSE_OK) {
+          callback(ledger::Result::LEDGER_ERROR);
+          return;
+        }
+
+        ledger_->ledger_client()->PublisherListNormalized(
```
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc:182]: It looks like this is a fix to make sure that `PublisherListNormalized` is called only after the database is updated. Is that correct?

File: vendor/bat-native-ledger/include/bat/ledger/ledger_client.h [hunk 10]
```diff
@@ -78,13 +78,15 @@ class LEDGER_EXPORT LedgerClient {
 
   virtual void LoadPublisherState(OnLoadCallback callback) = 0;
 
-  virtual void OnPanelPublisherInfo(Result result,
-                                   ledger::PublisherInfoPtr publisher_info,
-                                   uint64_t windowId) = 0;
+  virtual void OnPanelPublisherInfo(
+      Result result,
```
Comment (emerick) [inline vendor/bat-native-ledger/include/bat/ledger/ledger_client.h:82]: Maybe add a `ledger` namespace to this for consistency with other parameters.

File: vendor/bat-native-ledger/src/bat/ledger/internal/media/media.cc [hunk 11]
```diff
@@ -54,10 +54,13 @@ std::string Media::GetLinkType(
...
 
-void Media::ProcessMedia(const std::map<std::string, std::string>& parts,
-                               const std::string& type,
-                               ledger::VisitDataPtr visit_data) {
-  if (parts.size() == 0 || !ledger_->GetRewardsMainEnabled() || !visit_data) {
+void Media::ProcessMedia(
+    const std::map<std::string, std::string>& parts,
+    const std::string& type,
+    ledger::VisitDataPtr visit_data) {
+  if (parts.size() == 0 ||
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/internal/media/media.cc:61]: Checking for `parts.empty()` is more concise.

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.h (hunk 5 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/state/state.h:80]: removed

File: vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold.cc (hunk 6 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/uphold/uphold.cc:220]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc (hunk 7 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc:204]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc (hunk 8 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/state/state.cc:174]: fixed

File: vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc (hunk 9 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/database/database_activity_info.cc:182]: yes as values will only change if db is successfully updated and only in this case we need to normalize based on new values

File: vendor/bat-native-ledger/include/bat/ledger/ledger_client.h (hunk 10 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/include/bat/ledger/ledger_client.h:82]: added

File: vendor/bat-native-ledger/src/bat/ledger/internal/media/media.cc (hunk 11 above)
Comment (NejcZdovc) [reply vendor/bat-native-ledger/src/bat/ledger/internal/media/media.cc]: done

**Comment by NejcZdovc** [comment]
CI failed on known things
